### PR TITLE
monosat: Fix Linux build

### PR DESCRIPTION
**Pull request by acairncross**
###### Motivation for this change
The Linux build was broken because the CMake build process for this package builds both static and shared libraries, but the static library dependencies weren't being provided. For some reason this doesn't seem to be a problem for the Darwin build:
Linux: https://hydra.nixos.org/job/nixpkgs/trunk/monosat.x86_64-linux
Darwin: https://hydra.nixos.org/job/nixpkgs/trunk/monosat.x86_64-darwin

The closure size seems large compared to the Darwin build. Is this normal?
`/nix/store/1n2vwa7hc2wr882szg8s4dcajqc46m5b-monosat-09yhym2      511.3M`

~~I'm not sure how to build the accompanying Python package.~~

@copumpkin 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by acairncross** [comment]
Ah, the Python package (`python37Packages.monosat`) build successfully too.

**Comment by acairncross** [comment]
And maybe the Darwin/Linux closure size difference is because the Darwin build doesn't `includeJava`?

**Comment by Taneb** [comment]
@GrahamcOfBorg build monosat python37Packages.monosat

File: pkgs/applications/science/logic/monosat/default.nix [hunk 1]
```diff
@@ -20,10 +20,12 @@ let
     inherit rev sha256;
   };
 
+  # The CMake build is set up to build both shared and static libraries, so
+  # both shared and static libraries are needed as dependencies.
   core = stdenv.mkDerivation rec {
     name = "${pname}-${version}";
     inherit src;
-    buildInputs = [ cmake zlib gmp jdk8 ];
+    buildInputs = [ cmake zlib zlib.static gmp-static jdk8 git ];
```
Comment (JohnAZoidberg) [inline pkgs/applications/science/logic/monosat/default.nix]: It looks like you've removed the dynamic gmp dependency. Does the dynamic library still get built?

Comment (acairncross) [reply pkgs/applications/science/logic/monosat/default.nix]: Unlike zlib, which has its dynamic and static libraries separated into the `zlib` and `zlib.static` derivations, gmp seems to have a dynamic derivation and a dynamic+static derivation (built by setting the `withStatic` argument to true), so yes the dynamic library does get built.

**Comment by copumpkin** [comment]
Thanks! I'll try to test this over the weekend

**Comment by worldofpeace** [comment]
Can we disable static builds and just have it shared? `setup.sh` already has it set to do `--disable-static` so that's a guideline for other packages.

**Comment by acairncross** [comment]
I don't think that's straightforward to do. I think there would need to be changes made upstream to provide separate shared/static builds. I've made an issue on the monosat repo asking about this.

**Comment by acairncross** [comment]
I've made a PR on the monosat repo to add an option to do a dynamically linked build only. If that gets merged I should be able to use that option for the Nix package. Would it not make sense to have that as a separate PR from this, since this is just fixing a currently broken build, whereas removing the static build is more of an improvement?

**Comment by worldofpeace** [comment]
Your patch seems fine @acairncross. Can we use https://github.com/sambayless/monosat/pull/24 downstream here?

**Comment by worldofpeace** [comment]
Also, just for clarity, the build fails because it needs static libraries, which isn't really something we should curate in nixpkgs unless it's for the `pkgsStatic` overlay. So I'd think disabling them in anyway would be the most idiomatic fix to the build. Although in this case it required an improvement upstream.

**Comment by acairncross** [comment]
My patch was merged upstream so I've bumped up the version here to use it. Is that ok?

**Comment by worldofpeace** [comment]
The [diff](https://github.com/sambayless/monosat/compare/2deeadeff214e975c9f7508bc8a24fa05a1a0c32..60528a380a2fd17e84d04e86f14b2349e8f9fa51) looks, umm... Huge? You should upgrade it to the beta release
* https://github.com/sambayless/monosat/releases/tag/1.8.0

have that in your commit message
```
monosat: $old_version -> $new_version
```

and apply that patch on top of it.

As you're the only (and new) maintainer of this I'm fine deferring to you as to what's acceptable.
Currently it's broken so it doesn't matter greatly.

Though I will mention, the preferred practice would be to make a derivative of the patch upstream to apply to this version... but as stated :smile:

**Comment by worldofpeace** [comment]
Edit: it appears the commit is in the 1.8.0 tag https://github.com/sambayless/monosat/commit/033ac0fe9acc60a8513bbe40861114a7260d4612

**Comment by acairncross** [comment]
> The diff looks, umm... Huge?

Definitely a huge diff, although I think the majority of it is from applying a code formatter!

> the preferred practice would be to make a derivative of the patch upstream to apply to this version

Ah, ok!

Anyway, I've bumped it to 1.8.0 (and also squashed the commit that did the dynamic+static build).

File: pkgs/applications/science/logic/monosat/default.nix [hunk 2]
```diff
@@ -23,7 +23,7 @@ let
   core = stdenv.mkDerivation {
     name = "${pname}-${version}";
     inherit src;
-    buildInputs = [ cmake zlib gmp jdk8 ];
+    buildInputs = [ cmake zlib gmp jdk8 git ];
```
Comment (worldofpeace) [inline pkgs/applications/science/logic/monosat/default.nix]: git shouldn't be needed here
* https://github.com/sambayless/monosat/blob/fc2d4adc51b10150473b79646894aee687b6e310/CMakeLists.txt#L22

A flag `-DSHOW_GIT_VERSION=OFF` could be used.

And we fetch from an archive so `.git` isn't available so it doesn't do anything
* https://github.com/sambayless/monosat/blob/fc2d4adc51b10150473b79646894aee687b6e310/CMakeLists.txt#L345

Comment (acairncross) [reply pkgs/applications/science/logic/monosat/default.nix]: You're right. I thought I remember seeing error messages from git not being installed at some point, but I must have been doing something different when that happened.

I've removed the commit that adds git.

**Comment by worldofpeace** [comment]
Thanks @acairncross for the fix and the patience :sparkles: . 
Luckily @Taneb brought this to my attention so we could get this merged.

**Pull request merged**